core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -495,13 +495,39 @@ private[spark] class TaskSetManager(
    * Get the level we can launch tasks according to delay scheduling, based on current wait time.
    */
   private def getAllowedLocalityLevel(curTime: Long): TaskLocality.TaskLocality = {
-    while (curTime - lastLaunchTime >= localityWaits(currentLocalityIndex) &&
-        currentLocalityIndex < myLocalityLevels.length - 1)
-    {
-      // Jump to the next locality level, and remove our waiting time for the current one since
-      // we don't want to count it again on the next one
-      lastLaunchTime += localityWaits(currentLocalityIndex)
-      currentLocalityIndex += 1
+    // remove the emptyList from pendingTasks lazily
+    def hasNonEmptyList(pendingTasks: HashMap[String, ArrayBuffer[Int]]): Boolean = {
```

Comment:
If you have N executors, and they become empty one at a time, then this is a N^2 operation right (because you iterate over all of them -- not just the non-empty ones -- each time)?